walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
+   *         Returns empty if duration is greater than a day.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (events.isEmpty()) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return new ArrayList<>();
+      } else {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> attendeesWithOptional = new ArrayList<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return new ArrayList<>();
+    }
+
+    long duration = request.getDuration();
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      if (hasAttendeesAttending(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+
+      if (hasAttendeesAttending(event.getAttendees(), attendeesWithOptional)) {
+        occupiedTimeRangeWithOptional.add(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    if (!availableMeetings.isEmpty()) {
+      return availableMeetings;
+    } else {
+      // No mandatory attendees.
+      if (attendees.isEmpty()) {
+        return new ArrayList<>();
+      } else {
+        return getAvailableTimeRange(occupiedTimeRange, duration);
+      }
+    }
+  }
+
+  private boolean hasAttendeesAttending(Collection<String> eventAttendees, Collection<String> requestAttendees) {
+    for (String attendee: requestAttendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private Collection<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    Collection<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return availableMeetings;
+    }
+
+    if (occupiedTimeRange.isEmpty()) {
+      availableMeetings.add(TimeRange.WHOLE_DAY);
+      return availableMeetings;
+    }
+
+    List<TimeRange> mergedTimeRanges = mergeTimeRange(occupiedTimeRange);
+
+    int start = TimeRange.START_OF_DAY;
+
+    for (TimeRange timeRange: mergedTimeRanges) {
+      // Interval duration is not enough.
+      if ((timeRange.start() - start) < duration) {
+        int startForInclusive = Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY);
+        int startForNonInclusive = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+        start = timeRange.contains(timeRange.end()) ? startForInclusive : startForNonInclusive;
```

Comment:
With current implementation of timeRange:
timeRange.contains(timeRange.end()) is always false
Is it intended?

Reply:
As discussed offline timeRange.contains(timeRange.end()) is false

Reply:
Fixed